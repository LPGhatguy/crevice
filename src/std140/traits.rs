use std::io::{self, Write};
use std::mem::size_of;

use bytemuck::{bytes_of, Pod, Zeroable};

use crate::std140::Writer;

/// Trait implemented for all `std140` primitives. Generally should not be
/// implemented outside this crate.
pub unsafe trait Std140: Copy + Zeroable + Pod {
    /// The required alignment of the type. Must be a power of two.
    ///
    /// This is distinct from the value returned by `std::mem::align_of` because
    /// `AsStd140` structs do not use Rust's alignment. This enables them to
    /// control and zero their padding bytes, making converting them to and from
    /// slices safe.
    const ALIGNMENT: usize;

    /// Casts the type to a byte array. Implementors should not override this
    /// method.
    ///
    /// # Safety
    /// This is always safe due to the requirements of [`bytemuck::Pod`] being a
    /// prerequisite for this trait.
    fn as_bytes(&self) -> &[u8] {
        bytes_of(self)
    }
}

/**
Trait implemented for all types that can be turned into `std140` values.

This trait can often be `#[derive]`'d instead of manually implementing it. Any
struct which contains only fields that also implement `AsStd140` can derive
`AsStd140`.

Types from the mint crate implement `AsStd140`, making them convenient for use
in uniform types. Most Rust geometry crates, like cgmath, nalgebra, and
ultraviolet support mint.

## Example

```glsl
uniform CAMERA {
    mat4 view;
    mat4 projection;
} camera;
```

```
use cgmath::prelude::*;
use cgmath::{Matrix4, Deg, perspective};
use crevice::std140::{AsStd140, Std140};

#[derive(AsStd140)]
struct CameraUniform {
    view: mint::ColumnMatrix4<f32>,
    projection: mint::ColumnMatrix4<f32>,
}

let camera = CameraUniform {
    view: Matrix4::identity().into(),
    projection: perspective(Deg(60.0), 16.0/9.0, 0.01, 100.0).into(),
};

# fn write_to_gpu_buffer(bytes: &[u8]) {}
let camera_std140 = camera.as_std140();
write_to_gpu_buffer(camera_std140.as_bytes());
```
*/
pub trait AsStd140 {
    /// The `std140` version of this value.
    type Std140Type: Std140;

    /// Convert this value into the `std140` version of itself.
    fn as_std140(&self) -> Self::Std140Type;

    /// Returns the size of the `std140` version of this type. Useful for
    /// pre-sizing buffers.
    fn std140_size_static() -> usize {
        size_of::<Self::Std140Type>()
    }
}

impl<T> AsStd140 for T
where
    T: Std140,
{
    type Std140Type = Self;

    fn as_std140(&self) -> Self {
        *self
    }
}

/// Trait implemented for all types that can be written into a buffer as
/// `std140` bytes. This type is more general than [`AsStd140`]: all `AsStd140`
/// types implement `WriteStd140`, but not the other way around.
///
/// While `AsStd140` requires implementers to return a type that implements the
/// `Std140` trait, `WriteStd140` directly writes bytes using a [`Writer`]. This
/// makes `WriteStd140` usable for writing slices or other DSTs that could not
/// implement `AsStd140` without allocating new memory on the heap.
pub trait WriteStd140 {
    /// Writes this value into the given [`Writer`] using `std140` layout rules.
    ///
    /// Should return the offset of the first byte of this type, as returned by
    /// the first call to [`Writer::write`].
    fn write_std140<W: Write>(&self, writer: &mut Writer<W>) -> io::Result<usize>;

    /// The space required to write this value using `std140` layout rules. This
    /// does not include alignment padding that may be needed before or after
    /// this type when written as part of a larger buffer.
    fn std140_size(&self) -> usize {
        let mut writer = Writer::new(io::sink());
        self.write_std140(&mut writer).unwrap();
        writer.len()
    }
}

impl<T> WriteStd140 for T
where
    T: AsStd140,
{
    fn write_std140<W: Write>(&self, writer: &mut Writer<W>) -> io::Result<usize> {
        writer.write_std140(&self.as_std140())
    }

    fn std140_size(&self) -> usize {
        size_of::<<Self as AsStd140>::Std140Type>()
    }
}

impl<T> WriteStd140 for [T]
where
    T: WriteStd140,
{
    fn write_std140<W: Write>(&self, writer: &mut Writer<W>) -> io::Result<usize> {
        let mut offset = None;

        for item in self.iter() {
            let item_offset = item.write_std140(writer)?;

            if offset.is_none() {
                offset = Some(item_offset);
            }
        }

        Ok(offset.unwrap_or(writer.len()))
    }

    fn std140_size(&self) -> usize {
        let mut writer = Writer::new(io::sink());
        self.write_std140(&mut writer).unwrap();
        writer.len()
    }
}
