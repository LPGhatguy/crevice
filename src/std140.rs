//! Defines traits and types for working with data adhering to GLSL's `std140`
//! layout specification.

#[cfg(feature = "arrays")]
mod arrays;
mod dynamic_uniform;
mod primitives;
mod sizer;
mod traits;
#[cfg(feature = "std")]
mod writer;

pub use crate::bool::Bool;

#[cfg(feature = "arrays")]
pub use self::arrays::Std140ArrayItem;
pub use self::dynamic_uniform::*;
pub use self::primitives::*;
pub use self::sizer::*;
pub use self::traits::*;
#[cfg(feature = "std")]
pub use self::writer::*;

pub use crevice_derive::AsStd140;
