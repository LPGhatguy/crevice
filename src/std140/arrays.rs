use core::fmt::Debug;
use core::mem::{transmute_copy, MaybeUninit};

use bytemuck::{Pod, Zeroable};

use super::{AsStd140, Std140};

pub unsafe trait Std140ArrayItem: Std140 {
    type Padding: Zeroable + Copy + Debug;
}

type Padded<T> = (T, <T as Std140ArrayItem>::Padding);

fn wrap<T: AsStd140>(x: &T) -> Padded<T::Output>
where
    T::Output: Std140ArrayItem,
{
    (x.as_std140(), Zeroable::zeroed())
}

fn unwrap<T: AsStd140>(x: Padded<T::Output>) -> T
where
    T::Output: Std140ArrayItem,
{
    T::from_std140(x.0)
}

#[doc(hidden)]
#[derive(Copy, Clone, Debug)]
#[repr(transparent)]
pub struct Std140Array<T: Std140ArrayItem, const N: usize>([Padded<T>; N]);

unsafe impl<T: Std140ArrayItem, const N: usize> Zeroable for Std140Array<T, N> {}
unsafe impl<T: Std140ArrayItem, const N: usize> Pod for Std140Array<T, N> {}
unsafe impl<T: Std140ArrayItem, const N: usize> Std140 for Std140Array<T, N> {
    const ALIGNMENT: usize = crate::internal::max(16, T::ALIGNMENT);
}

impl<T: AsStd140, const N: usize> AsStd140 for [T; N]
where
    T::Output: Std140ArrayItem,
{
    type Output = Std140Array<T::Output, N>;
    fn as_std140(&self) -> Self::Output {
        let mut res: [MaybeUninit<_>; N] = unsafe { MaybeUninit::uninit().assume_init() };

        for i in 0..N {
            res[i] = MaybeUninit::new(wrap(&self[i]));
        }

        unsafe { transmute_copy(&res) }
    }

    fn from_std140(val: Self::Output) -> Self {
        val.0.map(|x| unwrap(x))
    }
}

unsafe impl Std140ArrayItem for f32 {
    type Padding = [u8; 12];
}

unsafe impl Std140ArrayItem for f64 {
    type Padding = [u8; 8];
}

unsafe impl Std140ArrayItem for i32 {
    type Padding = [u8; 12];
}

unsafe impl Std140ArrayItem for u32 {
    type Padding = [u8; 12];
}

unsafe impl Std140ArrayItem for crate::bool::Bool {
    type Padding = [u8; 12];
}
