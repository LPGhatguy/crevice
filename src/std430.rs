//! Defines traits and types for working with data adhering to GLSL's `std140`
//! layout specification.

mod primitives;
mod sizer;
mod traits;
mod writer;

pub use self::primitives::*;
pub use self::sizer::*;
pub use self::traits::*;
pub use self::writer::*;

pub use crevice_derive::AsStd430;
