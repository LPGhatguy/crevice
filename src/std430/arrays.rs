use core::fmt::Debug;
use core::mem::{transmute_copy, MaybeUninit};

use bytemuck::{Pod, Zeroable};

use super::{AsStd430, Std430};

pub unsafe trait Std430ArrayItem: Std430 {
    type Padding: Zeroable + Copy + Debug;
}

type Padded<T> = (T, <T as Std430ArrayItem>::Padding);

fn wrap<T: AsStd430>(x: &T) -> Padded<T::Output>
where
    T::Output: Std430ArrayItem,
{
    (x.as_std430(), Zeroable::zeroed())
}

fn unwrap<T: AsStd430>(x: Padded<T::Output>) -> T
where
    T::Output: Std430ArrayItem,
{
    T::from_std430(x.0)
}

#[doc(hidden)]
#[derive(Copy, Clone, Debug)]
#[repr(transparent)]
pub struct Std430Array<T: Std430ArrayItem, const N: usize>([Padded<T>; N]);

unsafe impl<T: Std430ArrayItem, const N: usize> Zeroable for Std430Array<T, N> {}
unsafe impl<T: Std430ArrayItem, const N: usize> Pod for Std430Array<T, N> {}
unsafe impl<T: Std430ArrayItem, const N: usize> Std430 for Std430Array<T, N> {
    const ALIGNMENT: usize = T::ALIGNMENT;
}

impl<T: AsStd430, const N: usize> AsStd430 for [T; N]
where
    T::Output: Std430ArrayItem,
{
    type Output = Std430Array<T::Output, N>;
    fn as_std430(&self) -> Self::Output {
        let mut res: [MaybeUninit<_>; N] = unsafe { MaybeUninit::uninit().assume_init() };

        for i in 0..N {
            res[i] = MaybeUninit::new(wrap(&self[i]));
        }

        unsafe { transmute_copy(&res) }
    }

    fn from_std430(val: Self::Output) -> Self {
        val.0.map(|x| unwrap(x))
    }
}

unsafe impl Std430ArrayItem for f32 {
    type Padding = [u8; 0];
}

unsafe impl Std430ArrayItem for f64 {
    type Padding = [u8; 0];
}

unsafe impl Std430ArrayItem for i32 {
    type Padding = [u8; 0];
}

unsafe impl Std430ArrayItem for u32 {
    type Padding = [u8; 0];
}

unsafe impl Std430ArrayItem for crate::bool::Bool {
    type Padding = [u8; 0];
}
