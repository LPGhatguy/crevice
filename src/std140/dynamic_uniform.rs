use bytemuck::{Pod, Zeroable};

use crate::internal::max;
use crate::std140::{AsStd140, Std140};

/// Wrapper type that aligns the inner type to at least 256 bytes.
///
/// This type is useful for ensuring correct alignment when creating dynamic
/// uniform buffers in APIs like WebGPU.
pub struct DynamicUniform<T>(pub T);

impl<T: AsStd140> AsStd140 for DynamicUniform<T> {
    type Output = DynamicUniformStd140<<T as AsStd140>::Output>;

    fn as_std140(&self) -> Self::Output {
        DynamicUniformStd140(self.0.as_std140())
    }

    fn from_std140(value: Self::Output) -> Self {
        DynamicUniform(<T as AsStd140>::from_std140(value.0))
    }
}

/// std140 variant of [`DynamicUniform`].
#[derive(Clone, Copy)]
#[repr(transparent)]
pub struct DynamicUniformStd140<T>(T);

unsafe impl<T: Std140> Std140 for DynamicUniformStd140<T> {
    const ALIGNMENT: usize = max(256, T::ALIGNMENT);
}

unsafe impl<T: Zeroable> Zeroable for DynamicUniformStd140<T> {}
unsafe impl<T: Pod> Pod for DynamicUniformStd140<T> {}

#[cfg(test)]
mod test {
    use super::*;

    use crate::std140::AsStd140;

    #[test]
    fn size_is_unchanged() {
        assert_eq!(
            DynamicUniform::<f32>::std140_size_static(),
            f32::std140_size_static()
        );
    }

    #[test]
    #[cfg(feature = "std")]
    fn alignment_applies() {
        use crate::std140;

        let mut output = Vec::new();
        let mut writer = std140::Writer::new(&mut output);

        writer.write(&DynamicUniform(0.0f32)).unwrap();
        assert_eq!(writer.len(), 4);

        writer.write(&DynamicUniform(1.0f32)).unwrap();
        assert_eq!(writer.len(), 260);
    }
}
