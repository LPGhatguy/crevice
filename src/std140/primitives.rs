#[cfg(feature = "arrays")]
use core::mem::size_of;

use bytemuck::{Pod, Zeroable};

use crate::bool::Bool;
use crate::glsl::Glsl;
#[cfg(feature = "arrays")]
use crate::internal::{align_offset, max};
use crate::std140::{AsStd140, Std140};

unsafe impl Std140 for f32 {
    const ALIGNMENT: usize = 4;
}

unsafe impl Std140 for f64 {
    const ALIGNMENT: usize = 8;
}

unsafe impl Std140 for i32 {
    const ALIGNMENT: usize = 4;
}

unsafe impl Std140 for u32 {
    const ALIGNMENT: usize = 4;
}

unsafe impl Std140 for Bool {
    const ALIGNMENT: usize = 4;
}

impl AsStd140 for bool {
    type Output = Bool;

    fn as_std140(&self) -> Self::Output {
        (*self).into()
    }

    fn from_std140(val: Self::Output) -> Self {
        val.into()
    }
}

macro_rules! vectors {
    (
        $(
            #[$doc:meta] align($align:literal) $glsl_name:ident $name:ident <$prim:ident> ($($field:ident),+)
        )+
    ) => {
        $(
            #[$doc]
            #[allow(missing_docs)]
            #[derive(Debug, Clone, Copy, PartialEq)]
            #[repr(C)]
            pub struct $name {
                $(pub $field: $prim,)+
            }

            unsafe impl Zeroable for $name {}
            unsafe impl Pod for $name {}

            unsafe impl Std140 for $name {
                const ALIGNMENT: usize = $align;
            }

            #[cfg(feature = "arrays")]
            unsafe impl super::Std140ArrayItem for $name {
                type Padding = [u8; align_offset(size_of::<$name>(), max(16, $align))];
            }

            unsafe impl Glsl for $name {
                const NAME: &'static str = stringify!($glsl_name);
            }
        )+
    };
}

vectors! {
    #[doc = "Corresponds to a GLSL `vec2` in std140 layout."] align(8) vec2 Vec2<f32>(x, y)
    #[doc = "Corresponds to a GLSL `vec3` in std140 layout."] align(16) vec3 Vec3<f32>(x, y, z)
    #[doc = "Corresponds to a GLSL `vec4` in std140 layout."] align(16) vec4 Vec4<f32>(x, y, z, w)

    #[doc = "Corresponds to a GLSL `ivec2` in std140 layout."] align(8) ivec2 IVec2<i32>(x, y)
    #[doc = "Corresponds to a GLSL `ivec3` in std140 layout."] align(16) ivec3 IVec3<i32>(x, y, z)
    #[doc = "Corresponds to a GLSL `ivec4` in std140 layout."] align(16) ivec4 IVec4<i32>(x, y, z, w)

    #[doc = "Corresponds to a GLSL `uvec2` in std140 layout."] align(8) uvec2 UVec2<u32>(x, y)
    #[doc = "Corresponds to a GLSL `uvec3` in std140 layout."] align(16) uvec3 UVec3<u32>(x, y, z)
    #[doc = "Corresponds to a GLSL `uvec4` in std140 layout."] align(16) uvec4 UVec4<u32>(x, y, z, w)

    #[doc = "Corresponds to a GLSL `bvec2` in std140 layout."] align(8) bvec2 BVec2<Bool>(x, y)
    #[doc = "Corresponds to a GLSL `bvec3` in std140 layout."] align(16) bvec3 BVec3<Bool>(x, y, z)
    #[doc = "Corresponds to a GLSL `bvec4` in std140 layout."] align(16) bvec4 BVec4<Bool>(x, y, z, w)

    #[doc = "Corresponds to a GLSL `dvec2` in std140 layout."] align(16) dvec2 DVec2<f64>(x, y)
    #[doc = "Corresponds to a GLSL `dvec3` in std140 layout."] align(32) dvec3 DVec3<f64>(x, y, z)
    #[doc = "Corresponds to a GLSL `dvec4` in std140 layout."] align(32) dvec4 DVec4<f64>(x, y, z, w)
}

macro_rules! matrices {
    (
        $(
            #[$doc:meta]
            align($align:literal)
            $glsl_name:ident $name:ident {
                $($field:ident: $field_ty:ty,)+
            }
        )+
    ) => {
        $(
            #[$doc]
            #[allow(missing_docs)]
            #[derive(Debug, Clone, Copy)]
            #[repr(C)]
            pub struct $name {
                $(pub $field: $field_ty,)+
            }

            unsafe impl Zeroable for $name {}
            unsafe impl Pod for $name {}

            unsafe impl Std140 for $name {
                const ALIGNMENT: usize = $align;
            }

            #[cfg(feature = "arrays")]
            unsafe impl super::Std140ArrayItem for $name {
                type Padding = [u8; 0];
            }

            unsafe impl Glsl for $name {
                const NAME: &'static str = stringify!($glsl_name);
            }
        )+
    };
}

matrices! {
    #[doc = "Corresponds to a GLSL `mat2` in std140 layout."]
    align(16)
    mat2 Mat2 {
        x: Vec2,
        _pad_x: [f32; 2],
        y: Vec2,
        _pad_y: [f32; 2],
    }

    #[doc = "Corresponds to a GLSL `mat3` in std140 layout."]
    align(16)
    mat3 Mat3 {
        x: Vec3,
        _pad_x: f32,
        y: Vec3,
        _pad_y: f32,
        z: Vec3,
        _pad_z: f32,
    }

    #[doc = "Corresponds to a GLSL `mat4` in std140 layout."]
    align(16)
    mat4 Mat4 {
        x: Vec4,
        y: Vec4,
        z: Vec4,
        w: Vec4,
    }

    #[doc = "Corresponds to a GLSL `dmat2` in std140 layout."]
    align(16)
    dmat2 DMat2 {
        x: DVec2,
        y: DVec2,
    }

    #[doc = "Corresponds to a GLSL `dmat3` in std140 layout."]
    align(32)
    dmat3 DMat3 {
        x: DVec3,
        _pad_x: f64,
        y: DVec3,
        _pad_y: f64,
        z: DVec3,
        _pad_z: f64,
    }

    #[doc = "Corresponds to a GLSL `dmat3` in std140 layout."]
    align(32)
    dmat4 DMat4 {
        x: DVec4,
        y: DVec4,
        z: DVec4,
        w: DVec4,
    }
}
