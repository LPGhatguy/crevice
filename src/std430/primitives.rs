use bytemuck::{Pod, Zeroable};

use crate::std430::Std430;

unsafe impl Std430 for f32 {
    const ALIGNMENT: usize = 4;
}

unsafe impl Std430 for f64 {
    const ALIGNMENT: usize = 8;
}

unsafe impl Std430 for i32 {
    const ALIGNMENT: usize = 4;
}

unsafe impl Std430 for u32 {
    const ALIGNMENT: usize = 4;
}

macro_rules! vectors {
    (
        $(
            #[$doc:meta] align($align:literal) $name:ident <$prim:ident> ($($field:ident),+)
        )+
    ) => {
        $(
            #[$doc]
            #[allow(missing_docs)]
            #[derive(Debug, Clone, Copy)]
            pub struct $name {
                $(pub $field: $prim,)+
            }

            unsafe impl Zeroable for $name {}
            unsafe impl Pod for $name {}

            unsafe impl Std430 for $name {
                const ALIGNMENT: usize = $align;
            }
        )+
    };
}

vectors! {
    #[doc = "Corresponds to a GLSL `vec2` in std430 layout."] align(8) Vec2<f32>(x, y)
    #[doc = "Corresponds to a GLSL `vec3` in std430 layout."] align(16) Vec3<f32>(x, y, z)
    #[doc = "Corresponds to a GLSL `vec4` in std430 layout."] align(16) Vec4<f32>(x, y, z, w)

    #[doc = "Corresponds to a GLSL `dvec2` in std430 layout."] align(16) DVec2<f64>(x, y)
    #[doc = "Corresponds to a GLSL `dvec3` in std430 layout."] align(32) DVec3<f64>(x, y, z)
    #[doc = "Corresponds to a GLSL `dvec4` in std430 layout."] align(32) DVec4<f64>(x, y, z, w)
}

macro_rules! matrices {
    (
        $(
            #[$doc:meta]
            align($align:literal)
            $name:ident {
                $($field:ident: $field_ty:ty,)+
            }
        )+
    ) => {
        $(
            #[$doc]
            #[allow(missing_docs)]
            #[derive(Debug, Clone, Copy)]
            pub struct $name {
                $(pub $field: $field_ty,)+
            }

            unsafe impl Zeroable for $name {}
            unsafe impl Pod for $name {}

            unsafe impl Std430 for $name {
                const ALIGNMENT: usize = $align;
            }
        )+
    };
}

matrices! {
    #[doc = "Corresponds to a GLSL `mat2` in std430 layout."]
    align(16)
    Mat2 {
        x: Vec2,
        y: Vec2,
    }

    #[doc = "Corresponds to a GLSL `mat3` in std430 layout."]
    align(16)
    Mat3 {
        x: Vec3,
        y: Vec3,
        z: Vec3,
    }

    #[doc = "Corresponds to a GLSL `mat4` in std430 layout."]
    align(16)
    Mat4 {
        x: Vec4,
        y: Vec4,
        z: Vec4,
        w: Vec4,
    }

    #[doc = "Corresponds to a GLSL `dmat2` in std430 layout."]
    align(16)
    DMat2 {
        x: DVec2,
        y: DVec2,
    }

    #[doc = "Corresponds to a GLSL `dmat3` in std430 layout."]
    align(32)
    DMat3 {
        x: DVec3,
        y: DVec3,
        z: DVec3,
    }

    #[doc = "Corresponds to a GLSL `dmat3` in std430 layout."]
    align(32)
    DMat4 {
        x: DVec4,
        y: DVec4,
        z: DVec4,
        w: DVec4,
    }
}
