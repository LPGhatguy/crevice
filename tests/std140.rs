use insta::assert_yaml_snapshot;
use type_layout::TypeLayout;

use crevice::std140::{AsStd140, DVec4, Std140, Vec3};
use crevice::std430::AsStd430;

#[derive(AsStd140)]
struct PrimitiveF32 {
    x: f32,
    y: f32,
}

#[test]
fn primitive_f32() {
    assert_yaml_snapshot!(<<PrimitiveF32 as AsStd140>::Std140Type as TypeLayout>::type_layout());

    assert_eq!(<PrimitiveF32 as AsStd140>::Std140Type::ALIGNMENT, 16);

    let value = PrimitiveF32 { x: 1.0, y: 2.0 };
    let _value_std140 = value.as_std140();
}

#[derive(AsStd140)]
struct TestVec3 {
    pos: Vec3,
    velocity: Vec3,
}

#[test]
fn test_vec3() {
    assert_yaml_snapshot!(<<TestVec3 as AsStd140>::Std140Type as TypeLayout>::type_layout());

    assert_eq!(<TestVec3 as AsStd140>::Std140Type::ALIGNMENT, 16);

    let value = TestVec3 {
        pos: Vec3 {
            x: 1.0,
            y: 2.0,
            z: 3.0,
        },
        velocity: Vec3 {
            x: 4.0,
            y: 5.0,
            z: 6.0,
        },
    };
    let _value_std140 = value.as_std140();
}

#[derive(AsStd140)]
struct UsingVec3Padding {
    pos: Vec3,
    brightness: f32,
}

#[test]
fn using_vec3_padding() {
    assert_yaml_snapshot!(
        <<UsingVec3Padding as AsStd140>::Std140Type as TypeLayout>::type_layout()
    );

    assert_eq!(<UsingVec3Padding as AsStd140>::Std140Type::ALIGNMENT, 16);

    let value = UsingVec3Padding {
        pos: Vec3 {
            x: 1.0,
            y: 2.0,
            z: 3.0,
        },
        brightness: 4.0,
    };
    let _value_std140 = value.as_std140();
}

#[derive(AsStd140)]
struct PointLight {
    position: Vec3,
    diffuse: Vec3,
    specular: Vec3,
    brightness: f32,
}

#[test]
fn point_light() {
    assert_yaml_snapshot!(<<PointLight as AsStd140>::Std140Type as TypeLayout>::type_layout());

    assert_eq!(<PointLight as AsStd140>::Std140Type::ALIGNMENT, 16);

    let value = PointLight {
        position: Vec3 {
            x: 1.0,
            y: 2.0,
            z: 3.0,
        },
        diffuse: Vec3 {
            x: 1.0,
            y: 2.0,
            z: 3.0,
        },
        specular: Vec3 {
            x: 1.0,
            y: 2.0,
            z: 3.0,
        },
        brightness: 4.0,
    };
    let _value_std140 = value.as_std140();
}

#[derive(AsStd140)]
struct MoreThan16Alignment {
    doubles: DVec4,
}

#[test]
fn more_than_16_alignment() {
    assert_yaml_snapshot!(
        <<MoreThan16Alignment as AsStd140>::Std140Type as TypeLayout>::type_layout()
    );

    assert_eq!(<MoreThan16Alignment as AsStd140>::Std140Type::ALIGNMENT, 32);
}

#[derive(AsStd140)]
struct PaddingAtEnd {
    base_value: PrimitiveF32,
    small_field: f32,
}

#[test]
fn padding_at_end() {
    assert_yaml_snapshot!(<<PaddingAtEnd as AsStd140>::Std140Type as TypeLayout>::type_layout());
}

#[derive(AsStd140, AsStd430)]
struct MatrixUniform {
    e: mint::ColumnMatrix3<f32>,
    f: f32,
}

#[test]
fn matrix_uniform_std140() {
    assert_yaml_snapshot!(<<MatrixUniform as AsStd140>::Std140Type as TypeLayout>::type_layout());
}

#[test]
fn matrix_uniform_std430() {
    assert_yaml_snapshot!(<<MatrixUniform as AsStd430>::Std430Type as TypeLayout>::type_layout())
}

/// Rust size: 4, align: 4
/// Std140 size: 4, align: 16
#[derive(AsStd140)]
struct PaddedByStdButNotRust {
    x: f32,
}

/// Rust size: 8, align: 4
/// Std140 size: 20, align: 16
#[derive(AsStd140)]
struct BaseSizeAndStdSizeAreDifferent {
    first: PaddedByStdButNotRust,
    second: PaddedByStdButNotRust,
}

/// If checking for base struct size, produces layout:
/// (padding 0) (field 20) (padding 8) (field 4)
/// which does not properly align the second member.
#[derive(AsStd140)]
struct ProperlyChecksForUnderlyingTypeSize {
    leading: BaseSizeAndStdSizeAreDifferent,
    trailing: PaddedByStdButNotRust,
}

#[test]
fn proper_offset_calculations_for_differing_member_sizes() {
    assert_yaml_snapshot!(
        <<ProperlyChecksForUnderlyingTypeSize as AsStd140>::Std140Type as TypeLayout>::type_layout(
        )
    )
}

#[derive(AsStd140, Debug, PartialEq)]
struct ThereAndBackAgain {
    view: mint::ColumnMatrix3<f32>,
    origin: mint::Vector3<f32>,
}

#[test]
fn there_and_back_again() {
    let x = ThereAndBackAgain {
        view: mint::ColumnMatrix3 {
            x: mint::Vector3 {
                x: 1.0,
                y: 0.0,
                z: 0.0,
            },
            y: mint::Vector3 {
                x: 0.0,
                y: 1.0,
                z: 0.0,
            },
            z: mint::Vector3 {
                x: 0.0,
                y: 0.0,
                z: 1.0,
            },
        },
        origin: mint::Vector3 {
            x: 0.0,
            y: 1.0,
            z: 2.0,
        },
    };
    let x_as = x.as_std140();
    assert_eq!(<ThereAndBackAgain as AsStd140>::from_std140(x_as), x);
}
