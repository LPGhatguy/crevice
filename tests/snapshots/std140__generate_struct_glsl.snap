---
source: tests/std140.rs
expression: "TestGlsl::glsl_definition()"

---
struct TestGlsl {
	vec3 foo;
	mat2 bar;
};
